File: content/blog/maximum_twin_sum_of_a_linked_list/go/solution_half_stack.go
package maximum_twin_sum_of_a_linked_list

func pairSumStack(head *ListNode) int {
	stack := make([]int, 0)
	currentListElement, oddElementList := head, head

	for oddElementList != nil && oddElementList.Next != nil {
		stack = append(stack, currentListElement.Val)
		currentListElement = currentListElement.Next
		oddElementList = oddElementList.Next.Next
	}

	res := 0
	for ; currentListElement != nil; currentListElement = currentListElement.Next {
		res = max(res, currentListElement.Val+stack[len(stack)-1])
		stack = stack[:len(stack)-1]
	}

	return res
}

File: content/blog/maximum_twin_sum_of_a_linked_list/ts/solution_half_stack.ts
import { ListNode } from './ListNode';

export function pairSumStack(head: ListNode): number {
    let stack: number[] = [];
    let currentListElement: ListNode | null = head;
    let oddElementList: ListNode | null = head;

    // Заполняем стек первой половиной списка
    while (oddElementList !== null && oddElementList.next !== null) {
        stack.push(currentListElement!.val);
        currentListElement = currentListElement!.next;
        oddElementList = oddElementList.next.next;
    }

    let res = 0;

    // Сравниваем элементы с конца первой половины и из второй половины
    while (currentListElement !== null) {
        res = Math.max(res, currentListElement.val + stack.pop()!);
        currentListElement = currentListElement.next;
    }

    return res;
}